cpp/src/parquet/arrow/reader_internal.cc

```diff
@@ -713,6 +715,77 @@ Status TransferDecimal(RecordReader* reader, MemoryPool* pool,
   return Status::OK();
 }
 
+static inline Status ConvertToHalfFloat(const Array& array,
+                                        const std::shared_ptr<DataType>& type,
+                                        MemoryPool* pool, std::shared_ptr<Array>* out) {
+  constexpr int32_t byte_width = sizeof(uint16_t);
+  DCHECK_EQ(checked_cast<const ::arrow::HalfFloatType&>(*type).byte_width(), byte_width);
+
+  // We read the halffloat (uint16_t) bytes from a raw binary array, in which they're
+  // assumed to be little-endian.
+  const auto& binary_array = checked_cast<const ::arrow::FixedSizeBinaryArray&>(array);
+  DCHECK_EQ(checked_cast<const ::arrow::FixedSizeBinaryType&>(*binary_array.type())
+                .byte_width(),
+            byte_width);
+
+  // Number of elements in the halffloat array
+  const int64_t length = binary_array.length();
+  // Allocate data for the output halffloat array
+  ARROW_ASSIGN_OR_RAISE(auto data, ::arrow::AllocateBuffer(length * byte_width, pool));
+  uint8_t* out_ptr = data->mutable_data();
+
+  const int64_t null_count = binary_array.null_count();
+  // Copy the values to the output array in native-endian format
+  if (null_count > 0) {
+    for (int64_t i = 0; i < length; ++i, out_ptr += byte_width) {
+      Float16 f16{0};
```

Comment:
The code looks good to me, but why can't we blink memcpy when it's `LITTLE_ENDIAN`? Since if it's null, the value is undefined?

Reply:
Also, Parquet C++ has no big-endian compatibility currently.

Reply:
memcpy isn't even needed. You can (or should be able to) just reuse the buffers. For example by calling `array.View(type)`.